Some more changes to webapp-async (#40)
NLESC-JCER/run-cpp-on-web: js-webapp-async/README.md

# Introduction

> In an [earlier blogpost](../js-webapp/README.md) we discussed how to run c++ code on the web using Javascript. We created a webapp that executed some C++ code and then showed the result. While the page was running the C++ code, the page was blocked and unresponsive. That was not a problem then, because the computation done in the code was very quick. Blocked UI becomes a problem when we are performing tasks that take a bit longer to run. 

Reviewer: ```suggestion
In an [earlier blogpost](../js-webapp/README.md) we discussed how to run C++ code on the web using JavaScript. We created a web app that executed some C++ code and then showed the result. While the page was running the C++ code, the page was blocked and unresponsive. That was not a problem then, because the computation done in the code was very quick. Blocked UI becomes a problem when we are performing tasks that take a bit longer to run. 

```

> How to prevent blocking when running long running tasks in c++?

Reviewer: This sentence needs to be rephrased. Maybe combine with the next sentence?

Author: I agree that this could be phrased a bit smoother I guess. This is not the first iteration of this piece of text so it's kinda hard to improve it further. I guess for you, too, otherwise, you would have suggested something I suppose.

What specifically do you not like about this? Maybe even better to merge it like it is now, and improve this later.

In this blogpost, we will use web workers to solve this problem by running the code in another thread.

### Long-running tasks with web worker

> Let's have a look at the code we ended up with in last blog. When loading the page, the webassembly is excecuted, after which the page can finish rendering. Because the webassembly code was very quick, this was fine. For this blog we assume we have a longer running task. We create such task artificially by adding a few seconds of sleep in the [c++ code](newtonraphson.cpp). Like in the previous post, we compile the c++ code to create webassembly. The resulting page with our slow task is now [here](https://nlesc-jcer.github.io/run-cpp-on-web/js-webapp-async/example-blocking.html).

Reviewer: ```suggestion
Let's have a look at the code we ended up with in the last blog. When loading the page, the WebAssembly code is executed, after which the page can finish rendering. Because the WebAssembly code was very quick, this was fine. For this blog, we assume we have a longer running task. We create such task artificially by adding a few seconds of sleep in the [c++ code](newtonraphson.cpp). Like in the previous post, we compile the C++ code to create WebAssembly code. The example page with our slow task is can be found [here](https://nlesc-jcer.github.io/run-cpp-on-web/js-webapp-async/example-blocking.html).
```

> Notice that we also added a slider input to the page. This slider isn't connected to anything, and is there just to illustrate UI blocking. Notice that while the webassembly code is still running, the slider is completely unresponsive. If this were an actual webapp and not just a demo, the UI blocking would surely annoy any users of the system and possibly make working with the app cumbersome and impracticle. We can easily solve this, and keep the UI responsive at all times, using web workers.

Reviewer: ```suggestion
Notice that we also added a slider input to the page. This slider has no function, and is there just to illustrate UI blockage. Notice that while the WebAssembly code is still running, the slider is completely unresponsive. If this was an actual web app and not just a demo, the UI blocking would surely annoy users and possibly make working with the app cumbersome and impractical. We can easily solve this, and keep the UI responsive at all times, using web workers.

```

![blocking ui](blocking.gif)

Blocked ui while code is running

### Web workers

> A [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) is an object that handles execution of a piece of code in another thread. On completion the web worker will notify so the result can be processed and, in this case, rendered on the page. 

Reviewer: ```suggestion
A [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API) is an object that handles execution of a piece of code in another thread. On completion the web worker will send a notification so the result can be processed and, in this case, rendered on the page. 

```

> The way the page communicates with the worker object is through sending messages. The page will send a message to the worker to start doing work, and the message will include all data that the worder needs. The worker will perform its execution, using only the data that was in the message, as the worker will not be able to access any data in the webapp directly. When finished the worker needs to communicate the results back to the webapp. It can do this by sending a message. 

Reviewer: ```suggestion
The way the page communicates with the worker object is through sending messages. The page will send a message to the worker to start doing work, and the message will include all data that the worker needs. The worker will perform its execution, using only the data that was in the message, as the worker will not be able to access any data in the web app directly. When finished the worker needs to communicate the results back to the web app. It will do this by sending a message. 

```

The webpage that uses the worker will look like:
```html
<!doctype html>
<html lang="en">
  <head>

>     <title>Example web worker</title>

Reviewer: ```suggestion
    <title>Web worker example</title>

```

    <script>
      const worker = new Worker('worker.js');
      worker.postMessage({
        type: 'CALCULATE',

>         payload: { epsilon: 0.001, guess: -20 }

Reviewer: ```suggestion
        payload: { epsilon: 0.001, guess: -4 }

```

Reviewer 2: Watch out this also requires changes in source code and wasm file, best to do in another PR

Author: I agree with Stefan and rather do this in one go.

> The worker code will contain only handling the incoming message. In this case, the worker will unpack the message, do the calculation, and pack the results in a new message that it will send back.

Reviewer: ```suggestion
The web worker code will contain only handling the incoming message. In this case, the web worker will unpack the message, do the calculation, and pack the results in a new message that it will send back.

```

    });
  }
};

```

We can see the code in action [here](https://nlesc-jcer.github.io/run-cpp-on-web/js-webapp-async/example-web-worker.html). The calculation still takes the same time to perform, but as you will notice, the slider will be responsive all the time.

![responsive ui](web-worker.gif)

Responsive ui thanks to web-worker

# Outro

> In this blogpost we have learned how to run, long running c++ code on the web using web workers.

Reviewer: ```suggestion
In this blog post we have learned how to run long running C++ code on the web using web workers.

```
